internal/services/rate.go

```diff
@@ -1,115 +1,160 @@
 package services
 
 import (
+	"context"
 	"encoding/json"
 	"errors"
+	"fmt"
 	"net/http"
-	"sync"
+	"strings"
 	"time"
-)
 
-const RatesAPIBaseURL = "https://cdn.jsdelivr.net/npm/@fawazahmed0/currency-api@latest/v1/currencies"
+	"github.com/fdemchenko/exchanger/internal/cache"
+	"github.com/rs/zerolog/log"
+)
 
 const (
-	RetryInterval = 500 * time.Millisecond
-	RetryCount    = 3
+	NBUExchangeRateURL        = "https://bank.gov.ua/NBUStatService/v1/statdirectory/exchange?json"
+	FawazAhmedExchangeRateURL = "https://cdn.jsdelivr.net/npm/@fawazahmed0/currency-api@latest/v1/currencies"
+
+	DefaultCachingDuration = 15 * time.Minute
+	RateFetchTimeout       = 10 * time.Second
 )
 
-type Rate struct {
+var ErrInvalidCurrencyCode = errors.New("invalid currency code")
+
+type ExchangeRateFetcher func(code string, client *http.Client) (float32, error)
+
+type NBUResponse struct {
+	Rate float32 `json:"rate"`
+	Code string  `json:"cc"`
+}
+
+func NBURateFetcher(code string, client *http.Client) (float32, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), RateFetchTimeout)
+	defer cancel()
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, NBUExchangeRateURL, nil)
+	if err != nil {
+		return 0, err
+	}
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return 0, err
+	}
+	defer resp.Body.Close()
+	var nbuResponse []NBUResponse
+	err = json.NewDecoder(resp.Body).Decode(&nbuResponse)
+	if err != nil {
+		return 0, err
+	}
+	for _, rate := range nbuResponse {
+		if strings.EqualFold(code, rate.Code) {
+			return rate.Rate, nil
+		}
+	}
+	return 0, ErrInvalidCurrencyCode
+}
+
+type FawazAhmedResponse struct {
 	Rates struct {
 		UAH float32 `json:"uah"`
 	} `json:"usd"`
 }
 
-var ErrNoFetchOccurred = errors.New("no fecth occurred yet")
+func FawazAhmedRateFetcher(code string, client *http.Client) (float32, error) {
```

Comment:
It's cool that you've extended rate fetcher with currency code.
However you could make it even better by parameterizing both `from` and `to` currencies